tests/libs/util/socket_helper.cpp

```diff
@@ -348,16 +579,63 @@ _stream_receiver_socket::post_async_receive()
     }
 }
 
+void
+_stream_receiver_socket::send_async_response(_In_z_ const char* message)
+{
+    // Send a message to the remote host using the sender socket.
+    std::vector<char> send_buffer(message, message + strlen(message));
+    WSABUF wsa_send_buffer{static_cast<ULONG>(send_buffer.size()), reinterpret_cast<char*>(send_buffer.data())};
+    uint32_t bytes_sent = 0;
+    overlapped.hEvent = WSACreateEvent();
+    int32_t error =
+        WSASend(accept_socket, &wsa_send_buffer, 1, reinterpret_cast<LPDWORD>(&bytes_sent), 0, &overlapped, NULL);
+    if (error != 0) {
+        int wsaerr = WSAGetLastError();
+        FAIL("send_async_response failed with " << wsaerr);
+    }
+}
+
+void
+_stream_receiver_socket::complete_async_send(int timeout_in_ms)
+{
+    int error = 0;
+    uint32_t bytes_sent = 0;
+    uint32_t send_flags = 0;
+    // Wait for the receiver socket to receive the message.
+    error = WSAWaitForMultipleEvents(1, &overlapped.hEvent, TRUE, timeout_in_ms, TRUE);
+    if (error == WSA_WAIT_EVENT_0) {
+
+        if (!WSAGetOverlappedResult(
+                socket,
+                &overlapped,
+                reinterpret_cast<LPDWORD>(&bytes_sent),
+                FALSE,
+                reinterpret_cast<LPDWORD>(&send_flags)))
+            FAIL("WSASend on the receiver socket failed with error: " << WSAGetLastError());
+        WSACloseEvent(overlapped.hEvent);
+        overlapped.hEvent = INVALID_HANDLE_VALUE;
+    } else {
+        FAIL("Waiting on receiver socket failed with " << error);
+    }
+}
+
 void
 _stream_receiver_socket::get_sender_address(_Out_ PSOCKADDR& from, _Out_ int& from_length)
 {
     from = (PSOCKADDR)(recv_buffer.data() + message_length);
     from_length = sizeof(sockaddr_storage);
 }
 
+void
+_stream_receiver_socket::get_local_address(_Out_ PSOCKADDR& from, _Out_ int& from_length)
+{
+    from = (PSOCKADDR)(recv_buffer.data() + message_length + sizeof(sockaddr_storage) + 16);
+    from_length = sizeof(sockaddr_storage);
```

Comment:
It would be better to return the actual length of the address rather than the full sockaddr.

Reply:
Removed this function now as not needed.